sound/soc/sof/intel/hda-loader-skl.c

```diff
@@ -106,30 +106,24 @@
 			(((x) << HDA_CL_SPBFIFO_SPBFCCTL_SPIBE_SHIFT) & \
 			 HDA_CL_SPBFIFO_SPBFCCTL_SPIBE_MASK)
 
-static int cl_skl_cldma_setup_bdle(struct snd_sof_dev *sdev,
-				   struct snd_dma_buffer *dmab_data,
-				   __le32 **bdlp, int size, int with_ioc)
+static void cl_skl_cldma_setup_bdle(struct snd_sof_dev *sdev,
+				    struct snd_dma_buffer *dmab_data,
+				    __le32 **bdlp, int size, int with_ioc)
 {
 	__le32 *bdl = *bdlp;
-	int frags = 0;
+	phys_addr_t addr = virt_to_phys(dmab_data->area);
 
-	while (size > 0) {
-		phys_addr_t addr = virt_to_phys(dmab_data->area +
-						(frags * size));
+	/* Set the address */
+	bdl[0] = cpu_to_le32(lower_32_bits(addr));
+	bdl[1] = cpu_to_le32(upper_32_bits(addr));
 
-		bdl[0] = cpu_to_le32(lower_32_bits(addr));
-		bdl[1] = cpu_to_le32(upper_32_bits(addr));
+	/* Set the DMA buffer size */
+	bdl[2] = cpu_to_le32(size);
 
-		bdl[2] = cpu_to_le32(size);
+	/* if the Interrupt on Complete is enabled */
+	bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
 
-		size -= size;
-		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
-
-		bdl += 4;
-		frags++;
-	}
-
-	return frags;
+	return;
```

Comment:
Humm, that seems too good to be true. Why would there be a single contiguous memory? I don't buy this, sorry

Reply:
@plbossart I keep this aligned with SST driver, I think this need a smaller continuous DMA hardware memory? (CLDMA need SNDRV_DMA_TYPE_DEV type continues buffer, stream DMA use SG buffer). Using a smaller DMA buffer and copy from the binary buffer multi times.
There are two solutions:
1. small DMA hardware buffer and copy to it multi times
`(file buffer) +--------+--------+--------+--`
`(memory copy) 1........2........3........4..`
`(DMA buffer ) +========+`

2. big DMA buffer and copy to it once
`(file buffer) +----------------------------+`
`(memory copy) 1.............................`
`(DMA buffer ) +========+========+========+========+`

In SST driver also use the solution 1:
`	ctx->cl_dev.ops.cl_setup_bdle(ctx, &ctx->cl_dev.dmab_data,
			&bdl, ctx->cl_dev.bufsize, 1);`

the 4th parameter of `size` is `ctx->cl_dev.bufsize`
`static void skl_cldma_setup_bdle(struct sst_dsp *ctx,
		struct snd_dma_buffer *dmab_data,
		__le32 **bdlp, int size, int with_ioc)`

then in skl_cldma_setup_bdle
`		size -= ctx->cl_dev.bufsize;`

which solution do you prefer?

Reply:
Please look at the following code, the assignments are done in a loop.

````c
/* Code loader helper APIs */
static void skl_cldma_setup_bdle(struct sst_dsp *ctx,
		struct snd_dma_buffer *dmab_data,
		__le32 **bdlp, int size, int with_ioc)
{
	__le32 *bdl = *bdlp;

	ctx->cl_dev.frags = 0;
	while (size > 0) {
		phys_addr_t addr = virt_to_phys(dmab_data->area +
				(ctx->cl_dev.frags * ctx->cl_dev.bufsize));

		bdl[0] = cpu_to_le32(lower_32_bits(addr));
		bdl[1] = cpu_to_le32(upper_32_bits(addr));

		bdl[2] = cpu_to_le32(ctx->cl_dev.bufsize);

		size -= ctx->cl_dev.bufsize;
		bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);

		bdl += 4;
		ctx->cl_dev.frags++;
	}
}
````

Reply:
@plbossart Yes, there is a loop in `skl_cldma_setup_bdle`, but the place where it is called give the `size` (the 4th parameter) as `ctx->cl_dev.bufsize`:
`	ctx->cl_dev.ops.cl_setup_bdle = skl_cldma_setup_bdle;`
`	ctx->cl_dev.ops.cl_setup_bdle(ctx, &ctx->cl_dev.dmab_data,
			&bdl, ctx->cl_dev.bufsize, 1);`
that make the loop loop only one time. The loops actually done is in `skl_cldma_copy_to_buf`.

What I am missing here?

Reply:
@plbossart

Reply:
@zhuyingjiang I went a couple of steps further - feel free to use my attached diff
[24.gitdiff.txt](https://github.com/thesofproject/linux/files/2911326/24.gitdiff.txt)

Reply:
@lyakh Thanks, update my PR. except that `bdl[3] = cpu_to_le32(0x01);` should be `bdl[3] = cpu_to_le32(0x0);`
@plbossart Is this what you want?

Reply:
@zhuyingjiang I don't think bdl[3] should be 0. In the original version:
```
bdl[3] = (size || !with_ioc) ? 0 : cpu_to_le32(0x01);
```
where size == 0, with_ioc = 1, i.e.
```
bdl[3] = (0 || !1) ? 0 : cpu_to_le32(0x01);
```
further
```
bdl[3] = 0 ? 0 : cpu_to_le32(0x01);
```
i.e.
```
bdl[3] = cpu_to_le32(0x01);
```

Reply:
@lyakh Oh, thanks, my mistake! updated.